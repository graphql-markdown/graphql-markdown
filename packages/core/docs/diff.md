# diff

## Schema

### hasChanges()

```ts
function hasChanges(
   schema, 
   tmpDir, 
   diffMethod, 
diffModule): Promise<boolean>
```

Defined in: [diff.ts:62](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/diff.ts#L62)

Determines if there are changes in the GraphQL schema by using a specified diff method and module.

#### Parameters

##### schema

`GraphQLSchema`

The GraphQL schema to check for changes.

##### tmpDir

`string`

The temporary directory to store intermediate files during the diff process.

##### diffMethod

`Maybe`\<`DiffMethodName`\>

The name of the diff method to use. Must be a string or `null`.

##### diffModule

`Maybe`\<`string`\> = `"@graphql-markdown/diff"`

The module to import for performing the diff. Defaults to `@graphql-markdown/diff`.

#### Returns

`Promise`\<`boolean`\>

A promise that resolves to `true` if changes are detected or if the diff method/module is invalid, otherwise `false`.

#### Examples

```typescript
import { hasChanges } from "./diff";
import { buildSchema } from "graphql";

const schema = buildSchema(`
  type Query {
    hello: String
  }
`);

const changesDetected = await hasChanges(schema, "/tmp", "methodName");
console.log(changesDetected); // true or false
```

```typescript
import { hasChanges } from "./diff";

const schema = getMySchema();
const result = await hasChanges(
  schema,
  "/tmp/schema-diff",
  "breaking",
  "./my-custom-diff-module"
);
```

#### Throws

Will log a warning if the specified diff module cannot be found.

#### See

 - DiffMethodName for available diff methods
 - FunctionCheckSchemaChanges for the signature of the function imported from the diff module

#### Since

1.0.0
