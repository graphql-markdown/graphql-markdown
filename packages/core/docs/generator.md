# generator

## Functions

### generateDocFromSchema()

```ts
function generateDocFromSchema(options): Promise<void>
```

Defined in: [generator.ts:66](https://github.com/graphql-markdown/graphql-markdown/blob/main/packages/core/src/generator.ts#L66)

Generates Markdown documentation from a GraphQL schema.

This function is the main entry point for the documentation generation process.
It loads the schema, analyzes it, and generates Markdown files according to the provided options.

#### Parameters

##### options

`GeneratorOptions`

Configuration options for the documentation generator

#### Returns

`Promise`\<`void`\>

Promise that resolves when documentation generation is complete

#### Example

```typescript
await generateDocFromSchema({
  baseURL: '/docs',
  outputDir: './docs',
  schemaLocation: './schema.graphql'
});
```
