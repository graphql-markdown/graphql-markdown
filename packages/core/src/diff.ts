import type {
  DiffMethodName,
  FunctionCheckSchemaChanges,
  GraphQLSchema,
  Maybe,
} from "@graphql-markdown/types";

import { log, LogLevel } from "@graphql-markdown/logger";

/**
 * Module for detecting changes in GraphQL schemas.
 * This module provides functionality to compare GraphQL schemas and detect changes.
 *
 * @module diff
 * @category Core
 * @since 1.0.0
 */

/**
 * Determines if there are changes in the GraphQL schema by using a specified diff method and module.
 *
 * @param schema - The GraphQL schema to check for changes.
 * @param tmpDir - The temporary directory to store intermediate files during the diff process.
 * @param diffMethod - The name of the diff method to use. Must be a string or `null`.
 * @param diffModule - The module to import for performing the diff. Defaults to `@graphql-markdown/diff`.
 * @returns A promise that resolves to `true` if changes are detected or if the diff method/module is invalid, otherwise `false`.
 *
 * @example
 * ```typescript
 * import { hasChanges } from "./diff";
 * import { buildSchema } from "graphql";
 *
 * const schema = buildSchema(`
 *   type Query {
 *     hello: String
 *   }
 * `);
 *
 * const changesDetected = await hasChanges(schema, "/tmp", "methodName");
 * console.log(changesDetected); // true or false
 * ```
 *
 * @example Using with a custom diff module
 * ```typescript
 * import { hasChanges } from "./diff";
 *
 * const schema = getMySchema();
 * const result = await hasChanges(
 *   schema,
 *   "/tmp/schema-diff",
 *   "breaking",
 *   "./my-custom-diff-module"
 * );
 * ```
 *
 * @throws Will log a warning if the specified diff module cannot be found.
 * @see {@link DiffMethodName} for available diff methods
 * @see {@link FunctionCheckSchemaChanges} for the signature of the function imported from the diff module
 * @category Schema
 * @since 1.0.0
 */
export const hasChanges = async (
  schema: GraphQLSchema,
  tmpDir: string,
  diffMethod: Maybe<DiffMethodName>,
  diffModule: Maybe<string> = "@graphql-markdown/diff",
): Promise<boolean> => {
  if (typeof diffMethod !== "string" || typeof diffModule !== "string") {
    return true;
  }

  try {
    const {
      checkSchemaChanges,
    }: { checkSchemaChanges: FunctionCheckSchemaChanges } = await import(
      diffModule
    );
    return await checkSchemaChanges(schema, tmpDir, diffMethod);
  } catch {
    log(
      `Cannot find module '${diffModule}' from @graphql-markdown/core!`,
      LogLevel.warn,
    );
    return true;
  }
};
