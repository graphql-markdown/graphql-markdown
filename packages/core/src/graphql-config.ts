/**
 * GraphQL Markdown configuration utilities
 *
 * This module provides utilities for loading and processing GraphQL configuration
 * using the graphql-config package.
 *
 * @module graphql-config
 */
import type {
  ClassName,
  ExtensionProjectConfig,
  GraphQLExtensionDeclaration,
  LoaderOption,
  Maybe,
  PackageConfig,
  PackageOptionsConfig,
} from "@graphql-markdown/types";

import { log } from "@graphql-markdown/logger";

/**
 * The name of the GraphQL Markdown extension.
 * Used to identify the extension in graphql-config.
 */
export const EXTENSION_NAME = "graphql-markdown" as const;

/**
 * GraphQL extension declaration for graphql-config.
 *
 * @returns The extension configuration object with name property.
 *
 * @example
 * ```typescript
 * // In graphql-config setup
 * const config = await loadConfig({
 *   extensions: [graphQLConfigExtension],
 * });
 * ```
 */
export const graphQLConfigExtension: GraphQLExtensionDeclaration = () => {
  return { name: EXTENSION_NAME } as const;
};

/**
 * Options for controlling throw behavior when loading configuration.
 *
 * @interface ThrowOptions
 * @property throwOnMissing Whether to throw when the config file is missing.
 * @property throwOnEmpty Whether to throw when the config is empty.
 */
interface ThrowOptions {
  throwOnMissing?: boolean;
  throwOnEmpty?: boolean;
}

/**
 * Sets loader options for GraphQL Markdown loaders.
 *
 * This function takes a LoaderOption object and merges the provided options
 * with any existing options for each loader.
 *
 * @param loaders - The loader configuration object.
 * @param options - The package options to apply to loaders.
 * @returns The updated loader configuration.
 *
 * @example
 * ```typescript
 * const loaders = {
 *   TypeScriptLoader: {
 *     module: "@graphql-markdown/typescript-loader",
 *     options: { baseDir: "./src" }
 *   }
 * };
 * const options = { outputDir: "./docs" };
 * const updatedLoaders = setLoaderOptions(loaders, options);
 * // Result: loaders with { baseDir: "./src", outputDir: "./docs" }
 * ```
 */
export const setLoaderOptions = (
  loaders: LoaderOption,
  options: PackageOptionsConfig,
): LoaderOption => {
  for (const loader in loaders) {
    if (typeof loaders[loader as ClassName] === "string") {
      loaders[loader as ClassName] = {
        module: loaders[loader as ClassName],
        options,
      } as PackageConfig;
    } else {
      (loaders[loader as ClassName] as PackageConfig).options = {
        ...options,
        ...(loaders[loader as ClassName] as PackageConfig).options,
      };
    }
  }
  return loaders;
};

/**
 * Loads the GraphQL Markdown configuration from graphql-config.
 *
 * This function attempts to load the GraphQL config and extract the
 * GraphQL Markdown extension configuration for the specified project ID.
 * It also normalizes schema configurations.
 *
 * @param id - The project ID to load configuration for.
 * @param options - Optional package options to apply.
 * @param throwOptions - Options for controlling throw behavior.
 * @returns The extension project configuration if found, otherwise undefined.
 *
 * @throws Will throw an error if throwOnMissing or throwOnEmpty is true and
 * the corresponding condition is met.
 *
 * @example
 * ```typescript
 * // Basic usage
 * const config = await loadConfiguration("my-project");
 *
 * // With options and throw behavior
 * const config = await loadConfiguration(
 *   "my-project",
 *   { baseDir: "./src" },
 *   { throwOnMissing: true, throwOnEmpty: false }
 * );
 * ```
 */
export const loadConfiguration = async (
  id: Maybe<string>,
  options?: Maybe<PackageOptionsConfig>,
  { throwOnMissing, throwOnEmpty }: ThrowOptions = {
    throwOnMissing: false,
    throwOnEmpty: false,
  },
): Promise<Maybe<Readonly<ExtensionProjectConfig>>> => {
  let graphQLConfig;

  if (typeof id !== "string") {
    return undefined;
  }

  try {
    graphQLConfig = await import("graphql-config");
  } catch {
    log("Cannot find module 'graphql-config'!");
    return undefined;
  }

  const config = await graphQLConfig.loadConfig({
    ...options,
    extensions: [graphQLConfigExtension],
    throwOnMissing,
    throwOnEmpty,
  });

  if (!config) {
    return undefined;
  }

  try {
    const projectConfig: ExtensionProjectConfig = config
      .getProject(id)
      .extension(EXTENSION_NAME);

    if (Array.isArray(projectConfig.schema)) {
      const schema = projectConfig.schema[0];
      if (typeof schema === "string") {
        projectConfig.schema = schema;
      }

      if (typeof projectConfig.schema === "object") {
        projectConfig.schema = Object.keys(schema)[0];

        if (typeof projectConfig.loaders !== "undefined") {
          projectConfig.loaders = setLoaderOptions(
            projectConfig.loaders ?? {},
            Object.values(schema)[0] as PackageOptionsConfig,
          );
        }
      }
    }

    return projectConfig as Readonly<ExtensionProjectConfig>;
  } catch {
    return undefined;
  }
};
