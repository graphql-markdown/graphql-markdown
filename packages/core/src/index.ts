/**
 * @primaryExport
 */
/* istanbul ignore file */

export { generateDocFromSchema } from "./generator";
export { buildConfig } from "./config";
