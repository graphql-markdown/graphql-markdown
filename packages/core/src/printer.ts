import type {
  IPrinter,
  Maybe,
  PackageName,
  Printer,
  PrinterConfig,
  PrinterOptions,
} from "@graphql-markdown/types";

export const getPrinter = async (
  printerModule?: Maybe<PackageName>,
  config?: Maybe<PrinterConfig>,
  options?: Maybe<PrinterOptions>,
  mdxParser?: PackageName,
): Promise<Printer> => {
  if (typeof printerModule !== "string") {
    throw new Error("Invalid printer module name.");
  }

  if (!config) {
    throw new Error("Invalid printer config.");
  }

  try {
    const { Printer }: { Printer: typeof IPrinter } = await import(
      printerModule
    );

    const { schema, baseURL, linkRoot } = config;
    await Printer.init(schema, baseURL, linkRoot, { ...options }, mdxParser);

    return Printer;
  } catch {
    throw new Error(`Cannot find module '${printerModule}'.`);
  }
};
