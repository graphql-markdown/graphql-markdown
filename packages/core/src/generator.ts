/**
 * @module
 * Core generator functionality for GraphQL Markdown.
 * This module contains the main entry point for generating documentation from GraphQL schemas.
 */
import type {
  DiffMethodName,
  DirectiveName,
  GeneratorOptions,
  GraphQLDirective,
  Maybe,
  SchemaEntity,
  TypeHierarchyObjectType,
} from "@graphql-markdown/types";

import Logger, { log, LogLevel } from "@graphql-markdown/logger";

import {
  getCustomDirectives,
  getDocumentLoaders,
  getGroups,
  getSchemaMap,
  loadSchema,
} from "@graphql-markdown/graphql";

import { toString } from "@graphql-markdown/utils";

import { DiffMethod } from "./config";
import { hasChanges } from "./diff";
import { getPrinter } from "./printer";
import type { Renderer } from "./renderer";
import { getRenderer } from "./renderer";

/**
 * Constant representing nanoseconds per second.
 *
 * This constant is used for high-precision time measurements and conversions
 * between seconds and nanoseconds.
 *
 */
const NS_PER_SEC = 1e9 as const;
/**
 * The number of decimal places to display when reporting times in seconds.
 *
 */
const SEC_DECIMALS = 3 as const;

/**
 * Generates Markdown documentation from a GraphQL schema.
 *
 * This function is the main entry point for the documentation generation process.
 * It loads the schema, analyzes it, and generates Markdown files according to the provided options.
 *
 * @param options - Configuration options for the documentation generator
 * @param options.baseURL - Base URL for the documentation links
 * @param options.customDirective - List of custom directives to include in documentation
 * @param options.diffMethod - Method to use for diffing schema changes (from {@link DiffMethodName})
 * @param options.docOptions - Documentation framework specific options
 * @param options.force - Forces regeneration of all files regardless of changes
 * @param options.groupByDirective - Directives used to group schema types
 * @param options.homepageLocation - Location of the homepage content
 * @param options.linkRoot - Root path for generating links
 * @param options.loadersList - List of loaders to use for loading the schema
 * @param options.loggerModule - Logger module to use
 * @param options.mdxParser - MDX parser module to use
 * @param options.metatags - Metadata tags to include in documentation
 * @param options.onlyDocDirective - Only document types with these directives
 * @param options.outputDir - Directory where documentation will be generated
 * @param options.prettify - Whether to prettify the generated output
 * @param options.printerModule - Printer module to use for generating output
 * @param options.printTypeOptions - Options for printing GraphQL types
 * @param options.printTypeOptions.deprecated - Whether to include deprecated types
 * @param options.printTypeOptions.hierarchy - Type hierarchy configuration
 * @param options.schemaLocation - Location of the GraphQL schema
 * @param options.skipDocDirective - Skip documenting types with these directives
 * @param options.tmpDir - Temporary directory for processing
 *
 * @returns Promise that resolves when documentation has been generated successfully. The function performs
 * schema analysis, generates markdown files for all schema entities, and prints a summary of the operation.
 *
 * @example
 * // Basic usage
 * await generateDocFromSchema({
 *   baseURL: "/docs/",
 *   schemaLocation: "./schema.graphql",
 *   outputDir: "./docs",
 *   diffMethod: DiffMethod.NONE,
 *   force: false,
 *   linkRoot: "/docs/",
 *   customDirective: [],
 *   groupByDirective: [],
 *   onlyDocDirective: [],
 *   skipDocDirective: [],
 *   loaders: ["@graphql-tools/json-file-loader"],
 *   printer: "@graphql-markdown/printer-docusaurus",
 *   printTypeOptions: {
 *     deprecated: true,
 *     hierarchy: { enabled: true }
 *   }
 * });
 *
 * @see {@link DiffMethod} - For available diff methods
 * @see {@link GeneratorOptions} - For all available configuration options
 * @see https://graphql-markdown.dev - For more examples and documentation
 */
export const generateDocFromSchema = async ({
  baseURL,
  customDirective,
  diffMethod,
  docOptions,
  force,
  groupByDirective,
  homepageLocation,
  linkRoot,
  loaders: loadersList,
  loggerModule,
  mdxParser,
  metatags,
  onlyDocDirective,
  outputDir,
  prettify,
  printer: printerModule,
  printTypeOptions,
  schemaLocation,
  skipDocDirective,
  tmpDir,
}: GeneratorOptions): Promise<void> => {
  const start = process.hrtime.bigint();

  await Logger(loggerModule);

  const loaders = await getDocumentLoaders(loadersList);

  if (!loaders) {
    log(
      `An error occurred while loading GraphQL loader.\nCheck your dependencies and configuration.`,
      "error",
    );
    return;
  }

  const schema = await loadSchema(schemaLocation as string, loaders);

  if (diffMethod !== DiffMethod.NONE) {
    const changed = await hasChanges(
      schema,
      tmpDir,
      diffMethod as DiffMethodName,
    );
    if (!changed) {
      log(`No changes detected in schema "${toString(schemaLocation)}".`);
    }
  }

  const [onlyDocDirectives, skipDocDirectives] = [
    onlyDocDirective,
    skipDocDirective,
  ].map((directiveList: DirectiveName[]) => {
    return directiveList
      .map((name: string) => {
        return schema.getDirective(name);
      })
      .filter((directive: Maybe<GraphQLDirective>) => {
        return typeof directive !== "undefined";
      }) as GraphQLDirective[];
  });

  const rootTypes = getSchemaMap(schema);
  const customDirectives = getCustomDirectives(rootTypes, customDirective);
  const groups = getGroups(rootTypes, groupByDirective);

  const mdxModule = await (!mdxParser
    ? undefined
    : import(mdxParser).catch(() => {
        log(
          `An error occurred while loading MDX formatter "${mdxParser}"`,
          LogLevel.warn,
        );
        return undefined;
      }));

  const printer = await getPrinter(
    // module mandatory
    printerModule,

    // config mandatory
    {
      baseURL,
      linkRoot,
      schema,
    },

    // options
    {
      customDirectives,
      groups,
      meta: {
        generatorFrameworkName: docOptions?.generatorFrameworkName,
        generatorFrameworkVersion: docOptions?.generatorFrameworkVersion,
      },
      metatags,
      onlyDocDirectives,
      printTypeOptions,
      skipDocDirectives,
    },
    mdxModule,
  );
  const renderer = await getRenderer(
    printer,
    outputDir,
    baseURL,
    groups,
    prettify,
    {
      ...docOptions,
      deprecated: printTypeOptions?.deprecated,
      force,
      hierarchy: printTypeOptions?.hierarchy as TypeHierarchyObjectType,
    },
    mdxModule,
  );

  const pages = await Promise.all(
    Object.keys(rootTypes).map(async (name) => {
      const typeName = name as SchemaEntity;
      return renderer.renderRootTypes(
        typeName,
        rootTypes[typeName],
      ) as ReturnType<Renderer["renderRootTypes"]>;
    }),
  );

  await renderer.renderHomepage(homepageLocation);

  const duration = (
    Number(process.hrtime.bigint() - start) / NS_PER_SEC
  ).toFixed(SEC_DECIMALS);

  log(
    `Documentation successfully generated in "${outputDir}" with base URL "${baseURL}".`,
    "success",
  );
  log(
    `${
      pages.flat().length
    } pages generated in ${duration}s from schema "${toString(schemaLocation)}".`,
  );
};
