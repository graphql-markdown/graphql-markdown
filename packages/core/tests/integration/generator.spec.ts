/* eslint-disable @typescript-eslint/no-base-to-string */
import path, { join } from "node:path";

import { vol } from "memfs";
jest.mock("fs");
jest.mock("node:fs/promises");

import type {
  ClassName,
  ConfigPrintTypeOptions,
  DirectiveName,
  GeneratorOptions,
  MDXString,
  PackageName,
  TypeDiffMethod,
} from "@graphql-markdown/types";

jest.mock("@graphql-markdown/printer-legacy");
import { Printer } from "@graphql-markdown/printer-legacy";

jest.mock("@graphql-markdown/diff");
import * as diff from "@graphql-markdown/diff";

jest.mock(
  "mdx-parser-mock",
  () => {
    return {
      generateIndexMetafile: (dirPath: string, category: string): string => {
        return path.join(dirPath, category).toLocaleLowerCase();
      },
    };
  },
  { virtual: true },
);

import { generateDocFromSchema } from "../../src/generator";
import {
  DEFAULT_OPTIONS,
  DeprecatedOption,
  DiffMethod,
  TypeHierarchy,
} from "../../src/config";

describe("renderer", () => {
  beforeEach(() => {
    // silent console
    jest.spyOn(global.console, "info").mockImplementation(() => {});
    jest.spyOn(global.console, "error").mockImplementation(() => {});

    jest.spyOn(Printer, "printType").mockImplementation((value) => {
      return value as MDXString;
    });

    vol.fromJSON({
      "/output": null,
      "/temp_test": null,
      "/assets/generated.md": "Dummy homepage for tweet.graphql",
    });
  });

  afterEach(() => {
    vol.reset();
    jest.restoreAllMocks();
    jest.resetAllMocks();
  });

  describe("generateDocFromSchema()", () => {
    test("generates Markdown document structure from GraphQL schema", async () => {
      expect.assertions(1);

      const config: GeneratorOptions = {
        baseURL: "graphql",
        diffMethod: DiffMethod.NONE,
        docOptions: {},
        homepageLocation: "/assets/generated.md",
        linkRoot: "docs",
        loaders: {
          ["GraphQLFileLoader" as ClassName]:
            "@graphql-tools/graphql-file-loader" as PackageName,
        },
        metatags: [],
        mdxParser: "mdx-parser-mock" as PackageName,
        onlyDocDirective: [],
        outputDir: "/output",
        prettify: false,
        printer: "@graphql-markdown/printer-legacy" as PackageName,
        printTypeOptions: {
          ...DEFAULT_OPTIONS.printTypeOptions,
          deprecated: DeprecatedOption.DEFAULT,
          hierarchy: TypeHierarchy.ENTITY,
        },
        schemaLocation: join(__dirname, "../__data__/tweet.graphql"),
        skipDocDirective: [],
        tmpDir: "/temp",
      };

      await generateDocFromSchema(config);

      expect(vol.toJSON(config.outputDir, undefined, true)).toMatchSnapshot();
    });

    test('outputs "no schema changed" message when called twice', async () => {
      expect.assertions(1);

      const logSpy = jest.spyOn(console, "info");

      const config: GeneratorOptions = {
        baseURL: "graphql",
        customDirective: undefined,
        diffMethod: "SCHEMA-HASH" as TypeDiffMethod,
        docOptions: {},
        homepageLocation: "/assets/generated.md",
        linkRoot: "docs",
        loaders: {
          ["GraphQLFileLoader" as ClassName]:
            "@graphql-tools/graphql-file-loader" as PackageName,
        },
        mdxParser: "mdx-parser-mock" as PackageName,
        metatags: [],
        onlyDocDirective: [],
        outputDir: "/output",
        prettify: false,
        printer: "@graphql-markdown/printer-legacy" as PackageName,
        printTypeOptions: {
          ...DEFAULT_OPTIONS.printTypeOptions,
          deprecated: DeprecatedOption.DEFAULT,
          hierarchy: TypeHierarchy.ENTITY,
        } as Required<ConfigPrintTypeOptions>,
        schemaLocation: join(__dirname, "../__data__/tweet.graphql"),
        skipDocDirective: [],
        tmpDir: "/temp",
      };

      jest.spyOn(diff, "checkSchemaChanges").mockResolvedValue(false);
      await generateDocFromSchema(config);

      expect(logSpy).toHaveBeenCalledWith(
        `No changes detected in schema "${String(config.schemaLocation)}".`,
      );
    });

    test("Markdown document structure from GraphQL schema is correct when using grouping", async () => {
      expect.assertions(2);

      const config: GeneratorOptions = {
        baseURL: "graphql",
        schemaLocation: join(
          __dirname,
          "../__data__/schema_with_grouping.graphql",
        ),
        diffMethod: DiffMethod.NONE,
        docOptions: {},
        groupByDirective: {
          directive: "doc" as DirectiveName,
          field: "category",
          fallback: "misc",
        },
        homepageLocation: "/assets/generated.md",
        linkRoot: "docs",
        loaders: {
          ["GraphQLFileLoader" as ClassName]:
            "@graphql-tools/graphql-file-loader" as PackageName,
        },
        mdxParser: "mdx-parser-mock" as PackageName,
        metatags: [],
        onlyDocDirective: [],
        outputDir: "/output",
        prettify: false,
        printer: "@graphql-markdown/printer-legacy" as PackageName,
        printTypeOptions: {
          ...DEFAULT_OPTIONS.printTypeOptions,
          deprecated: DeprecatedOption.DEFAULT,
          hierarchy: TypeHierarchy.ENTITY,
        },
        skipDocDirective: [],
        tmpDir: "/temp",
      };

      await generateDocFromSchema(config);

      expect(vol.toJSON(config.outputDir, undefined, true)).toMatchSnapshot();
      expect(vol.toJSON(config.tmpDir, undefined, true)).toMatchSnapshot();
    });

    test("Markdown document structure from GraphQL schema is correct when using api hierarchy", async () => {
      expect.assertions(2);

      const config: GeneratorOptions = {
        baseURL: "graphql",
        schemaLocation: join(
          __dirname,
          "../__data__/schema_with_grouping.graphql",
        ),
        diffMethod: DiffMethod.NONE,
        docOptions: {},
        homepageLocation: "/assets/generated.md",
        linkRoot: "docs",
        loaders: {
          ["GraphQLFileLoader" as ClassName]:
            "@graphql-tools/graphql-file-loader" as PackageName,
        },
        mdxParser: "mdx-parser-mock" as PackageName,
        metatags: [],
        onlyDocDirective: [],
        outputDir: "/output",
        prettify: false,
        printer: "@graphql-markdown/printer-legacy" as PackageName,
        printTypeOptions: {
          ...DEFAULT_OPTIONS.printTypeOptions,
          hierarchy: { [TypeHierarchy.API]: {} },
        },
        skipDocDirective: [],
        tmpDir: "/temp",
      };

      await generateDocFromSchema(config);

      expect(vol.toJSON(config.outputDir, undefined, true)).toMatchSnapshot();
      expect(vol.toJSON(config.tmpDir, undefined, true)).toMatchSnapshot();
    });

    test("Markdown document structure from GraphQL schema is correct when using flat hierarchy", async () => {
      expect.assertions(2);

      const config: GeneratorOptions = {
        baseURL: "graphql",
        schemaLocation: join(
          __dirname,
          "../__data__/schema_with_grouping.graphql",
        ),
        diffMethod: DiffMethod.NONE,
        docOptions: {},
        homepageLocation: "/assets/generated.md",
        linkRoot: "docs",
        loaders: {
          ["GraphQLFileLoader" as ClassName]:
            "@graphql-tools/graphql-file-loader" as PackageName,
        },
        metatags: [],
        onlyDocDirective: [],
        outputDir: "/output",
        prettify: false,
        printer: "@graphql-markdown/printer-legacy" as PackageName,
        printTypeOptions: {
          ...DEFAULT_OPTIONS.printTypeOptions,
          hierarchy: { [TypeHierarchy.FLAT]: {} },
        },
        skipDocDirective: [],
        tmpDir: "/temp",
      };

      await generateDocFromSchema(config);

      expect(vol.toJSON(config.outputDir, undefined, true)).toMatchSnapshot();
      expect(vol.toJSON(config.tmpDir, undefined, true)).toMatchSnapshot();
    });
  });
});
