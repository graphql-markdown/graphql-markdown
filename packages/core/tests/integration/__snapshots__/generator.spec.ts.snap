// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer generateDocFromSchema() Markdown document structure from GraphQL schema is correct when using api hierarchy 1`] = `
{
  "generated.md": "Dummy homepage for tweet.graphql",
  "operations/directives/include.mdx": "include",
  "operations/directives/skip.mdx": "skip",
  "operations/mutations/add-course.mdx": "add-course",
  "operations/mutations/drop-course.mdx": "drop-course",
  "operations/mutations/update-gpa.mdx": "update-gpa",
  "operations/mutations/update-grade.mdx": "update-grade",
  "operations/queries/all-courses.mdx": "all-courses",
  "operations/queries/course.mdx": "course",
  "operations/queries/gpa.mdx": "gpa",
  "operations/queries/math-courses.mdx": "math-courses",
  "operations/queries/science-courses.mdx": "science-courses",
  "operations/queries/search-role.mdx": "search-role",
  "operations/queries/un-weighted-gpa.mdx": "un-weighted-gpa",
  "operations/queries/weighted-gpa.mdx": "weighted-gpa",
  "types/directives/deprecated.mdx": "deprecated",
  "types/directives/doc.mdx": "doc",
  "types/directives/no-doc.mdx": "no-doc",
  "types/directives/one-of.mdx": "one-of",
  "types/directives/public.mdx": "public",
  "types/directives/specified-by.mdx": "specified-by",
  "types/enums/roles.mdx": "roles",
  "types/interfaces/record.mdx": "record",
  "types/objects/course.mdx": "course",
  "types/objects/department-information.mdx": "department-information",
  "types/objects/semester.mdx": "semester",
  "types/objects/unused-type.mdx": "unused-type",
  "types/scalars/boolean.mdx": "boolean",
  "types/scalars/date.mdx": "date",
  "types/scalars/email-address.mdx": "email-address",
  "types/scalars/id.mdx": "id",
  "types/scalars/int.mdx": "int",
  "types/scalars/phone-number.mdx": "phone-number",
  "types/scalars/string.mdx": "string",
  "types/scalars/time.mdx": "time",
  "types/scalars/url.mdx": "url",
}
`;

exports[`renderer generateDocFromSchema() Markdown document structure from GraphQL schema is correct when using api hierarchy 2`] = `{}`;

exports[`renderer generateDocFromSchema() Markdown document structure from GraphQL schema is correct when using flat hierarchy 1`] = `
{
  "add-course.md": "add-course",
  "all-courses.md": "all-courses",
  "boolean.md": "boolean",
  "course.md": "course",
  "date.md": "date",
  "department-information.md": "department-information",
  "deprecated.md": "deprecated",
  "doc.md": "doc",
  "drop-course.md": "drop-course",
  "email-address.md": "email-address",
  "generated.md": "Dummy homepage for tweet.graphql",
  "gpa.md": "gpa",
  "id.md": "id",
  "include.md": "include",
  "int.md": "int",
  "math-courses.md": "math-courses",
  "no-doc.md": "no-doc",
  "one-of.md": "one-of",
  "phone-number.md": "phone-number",
  "public.md": "public",
  "record.md": "record",
  "roles.md": "roles",
  "science-courses.md": "science-courses",
  "search-role.md": "search-role",
  "semester.md": "semester",
  "skip.md": "skip",
  "specified-by.md": "specified-by",
  "string.md": "string",
  "time.md": "time",
  "un-weighted-gpa.md": "un-weighted-gpa",
  "unused-type.md": "unused-type",
  "update-gpa.md": "update-gpa",
  "update-grade.md": "update-grade",
  "url.md": "url",
  "weighted-gpa.md": "weighted-gpa",
}
`;

exports[`renderer generateDocFromSchema() Markdown document structure from GraphQL schema is correct when using flat hierarchy 2`] = `{}`;

exports[`renderer generateDocFromSchema() Markdown document structure from GraphQL schema is correct when using grouping 1`] = `
{
  "course/mutations/add-course.mdx": "add-course",
  "course/mutations/drop-course.mdx": "drop-course",
  "course/queries/all-courses.mdx": "all-courses",
  "course/queries/course.mdx": "course",
  "course/queries/math-courses.mdx": "math-courses",
  "course/queries/science-courses.mdx": "science-courses",
  "generated.md": "Dummy homepage for tweet.graphql",
  "grade/mutations/update-gpa.mdx": "update-gpa",
  "grade/mutations/update-grade.mdx": "update-grade",
  "grade/queries/gpa.mdx": "gpa",
  "grade/queries/un-weighted-gpa.mdx": "un-weighted-gpa",
  "grade/queries/weighted-gpa.mdx": "weighted-gpa",
  "misc/directives/deprecated.mdx": "deprecated",
  "misc/directives/doc.mdx": "doc",
  "misc/directives/include.mdx": "include",
  "misc/directives/no-doc.mdx": "no-doc",
  "misc/directives/one-of.mdx": "one-of",
  "misc/directives/public.mdx": "public",
  "misc/directives/skip.mdx": "skip",
  "misc/directives/specified-by.mdx": "specified-by",
  "misc/enums/roles.mdx": "roles",
  "misc/interfaces/record.mdx": "record",
  "misc/objects/course.mdx": "course",
  "misc/objects/department-information.mdx": "department-information",
  "misc/objects/semester.mdx": "semester",
  "misc/objects/unused-type.mdx": "unused-type",
  "misc/queries/search-role.mdx": "search-role",
  "misc/scalars/boolean.mdx": "boolean",
  "misc/scalars/date.mdx": "date",
  "misc/scalars/email-address.mdx": "email-address",
  "misc/scalars/id.mdx": "id",
  "misc/scalars/int.mdx": "int",
  "misc/scalars/phone-number.mdx": "phone-number",
  "misc/scalars/string.mdx": "string",
  "misc/scalars/time.mdx": "time",
  "misc/scalars/url.mdx": "url",
}
`;

exports[`renderer generateDocFromSchema() Markdown document structure from GraphQL schema is correct when using grouping 2`] = `{}`;

exports[`renderer generateDocFromSchema() generates Markdown document structure from GraphQL schema 1`] = `
{
  "directives/deprecated.mdx": "deprecated",
  "directives/include.mdx": "include",
  "directives/one-of.mdx": "one-of",
  "directives/skip.mdx": "skip",
  "directives/specified-by.mdx": "specified-by",
  "enums/media-type.mdx": "media-type",
  "generated.md": "Dummy homepage for tweet.graphql",
  "inputs/tweet-content.mdx": "tweet-content",
  "interfaces/node.mdx": "node",
  "mutations/create-tweet.mdx": "create-tweet",
  "mutations/delete-tweet.mdx": "delete-tweet",
  "mutations/mark-tweet-read.mdx": "mark-tweet-read",
  "objects/media.mdx": "media",
  "objects/meta.mdx": "meta",
  "objects/notification.mdx": "notification",
  "objects/stat.mdx": "stat",
  "objects/tweet.mdx": "tweet",
  "objects/user.mdx": "user",
  "queries/notifications-meta.mdx": "notifications-meta",
  "queries/tweet.mdx": "tweet",
  "queries/tweets-meta.mdx": "tweets-meta",
  "queries/tweets.mdx": "tweets",
  "queries/user.mdx": "user",
  "queries/users.mdx": "users",
  "scalars/boolean.mdx": "boolean",
  "scalars/date.mdx": "date",
  "scalars/id.mdx": "id",
  "scalars/int.mdx": "int",
  "scalars/sri.mdx": "sri",
  "scalars/string.mdx": "string",
  "scalars/url.mdx": "url",
  "subscriptions/notifications.mdx": "notifications",
}
`;
